kmod/src/inode.c

```diff
@@ -1559,33 +1558,42 @@ int scoutfs_drop_inode(struct inode *inode)
 /*
  * Find orphan items and process each one.
  *
- * Runtime of this will be bounded by the number of orphans, which could
- * theoretically be very large. If that becomes a problem we might want to push
- * this work off to a thread.
- *
  * This only scans orphans for this node.  This will need to be covered by
  * the rest of node zone cleanup.
+ *
+ * Returns -ENOENT if no orphans exist, 0 if they exist but none are yet deletable,
+ * 1 if it actually deleted some, or other -ret if badness.
  */
-int scoutfs_scan_orphans(struct super_block *sb)
+int scoutfs_delete_orphans(struct super_block *sb,
+			 struct scoutfs_bloom_block *bb, unsigned int max)
 {
 	struct scoutfs_sb_info *sbi = SCOUTFS_SB(sb);
 	struct scoutfs_lock *lock = sbi->rid_lock;
 	struct scoutfs_key key;
 	struct scoutfs_key last;
+	bool deleted = false;
+	bool orphan_found = false;
+	__le64 last_seq;
 	int err = 0;
-	int ret;
+	int ret = 0;
 
-	trace_scoutfs_scan_orphans(sb);
+	trace_scoutfs_delete_orphans(sb, bb, max);
 
 	init_orphan_key(&key, sbi->rid, 0);
 	init_orphan_key(&last, sbi->rid, ~0ULL);
 
-	while (1) {
-		ret = scoutfs_item_next(sb, &key, &last, NULL, 0, lock);
+	while (deleted < max) {
```

Comment:
It looks like this'll always consume the full list because it's comparing the deleted bool to the max.  (And max is an unfortunate collision with the min/max macros.. maybe limit?).  I guess we want to max/limit to be applied to items scanned, not necessarily deleted, because we're worried about unmount hanging while this work is busy.  And undeletable inodes should be rare, so we're not too worried about constantly finding the entire front of the list undeletable.